constraint-solver/src/solver/backtracking.rs

```diff
@@ -0,0 +1,159 @@
+
+use super::{Error, Solver};
+
+/// The maximum number of possible assignments to try when backtracking.
+const MAX_BACKTRACKING_WIDTH: u64 = 1 << 12;
+
+pub struct Backtracker<'a, T: FieldElement, V> {
+    solver: &'a Solver<T, V>,
+}
+
+impl<'a, T: FieldElement, V> Backtracker<'a, T, V> {
+    pub fn new(solver: &'a Solver<T, V>) -> Self {
+        Backtracker { solver }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Hash + Eq + Display + Debug> Backtracker<'_, T, V> {
+    /// Returns a map of variable assignments that are unique in the constraint system.
+    /// Returns an error a contradiction was found.
+    pub fn get_unique_assignments(&self) -> Result<BTreeMap<V, T>, Error> {
+        log::debug!("Starting backtracking with maximum width {MAX_BACKTRACKING_WIDTH}");
+        let variable_sets = self.get_brute_force_candidates();
+
+        log::debug!(
+            "Found {} sets of variables with few possible assignments. Checking each set...",
+            variable_sets.len()
+        );
+
+        let unique_assignments = variable_sets
+            .iter()
+            .map(|assignment_candidates| self.find_unique_assignment(assignment_candidates))
+            // Might error out of a contradiction was found.
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            // Might return None if the assignment is not unique.
+            .flatten()
+            .collect::<Vec<_>>();
+
+        log::debug!(
+            "{} variable sets with unique assignments found",
+            unique_assignments.len()
+        );
+
+        let mut result = BTreeMap::new();
+        for (variable, value) in unique_assignments.iter().flatten() {
+            if let Some(old_value) = result.insert(variable.clone(), *value) {
+                if old_value != *value {
+                    // Two assignments contradict each other.
+                    return Err(Error::BacktrackingError);
+                }
+            }
+        }
+
+        log::debug!("Total assignments: {}", result.len());
+        for (variable, value) in &result {
+            log::trace!("  {variable} = {value}");
+        }
+
+        Ok(result)
+    }
+
+    /// Returns all unique sets of variables that appear together in an identity
+    /// (either in an algebraic constraint or in the same field of a bus interaction),
+    /// IF the number of possible assignments is less than `MAX_BACKTRACKING_WIDTH`.
+    fn get_brute_force_candidates(&self) -> Vec<BTreeSet<V>> {
+        self.solver
+            .constraint_system
+            .iter()
+            .map(|expression| {
+                expression
+                    .referenced_variables()
+                    .cloned()
+                    .collect::<BTreeSet<_>>()
+            })
+            .unique()
+            .filter(|variables| !variables.is_empty())
+            .filter(|variables| self.has_few_possible_assignments(variables.iter().cloned()))
+            .collect()
```

Comment:
Also no need to collect here

Reply:
True, but then I need to call collect when calling, to be able to log the size. I did that, to keep that function more general.

In practice, the size of this vector is in the thousands, so I think it's fine to materialize it.